### PR TITLE
Do not modify attributes hash in upsert! method

### PR DESCRIPTION
**Pull request by velveret**
Duplicate the hash before modification

https://github.com/ejholmes/restforce/issues/159

### CODE REVIEW

File: lib/restforce/concerns/api.rb [hunk 1]
```diff
@@ -360,9 +360,13 @@ def upsert(*args)
       # Raises an exception if an error is returned from Salesforce, including the 300
       # error returned if the external ID provided matches multiple records (in which
       # case the conflicting IDs can be found by looking at the response on the error)
-      def upsert!(sobject, field, attrs)
+      def upsert!(sobject, field, attrs_arg)
+        attrs = attrs_arg.dup
```
Comment (timrogers) [inline lib/restforce/concerns/api.rb]: Feels a bit ugly having the argument named `attrs_arg` - could we just do the following? Should work fine I think? Just tried and it worked fine for me playing in `irb`.

```ruby
def upsert!(sobject, field, attrs)
  attrs = attrs.dup
  # ...
end

File: lib/restforce/concerns/api.rb [hunk 2]
```diff
@@ -360,9 +360,13 @@ def upsert(*args)
...
         external_id =
           extract_case_insensitive_string_or_symbol_key_from_hash!(attrs, field)
+        if field.to_s != "Id" && (external_id.nil? || external_id.strip.empty?)
```
Comment (timrogers) [inline lib/restforce/concerns/api.rb]: I think I'd prefer to leave this out for now, or at least to do it in a separate PR to keep things clear and simple.

**Comment by timrogers** [comment]
💖

**Comment by velveret** [comment]
Amended the commit based on the suggestions. Thanks!

**Comment by timrogers** [comment]
Looks good :+1: Specs are currently failing in Travis for some reason... I'll merge this and look at that another time.

File: lib/restforce/concerns/api.rb (hunk 2 above)
Comment (velveret) [inline lib/restforce/concerns/api.rb]: Opened follow-up PR in https://github.com/ejholmes/restforce/pull/297
